sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/EFOShardSubscribersPool.java

```diff
@@ -108,22 +110,27 @@ class EFOShardSubscribersPool {
    * <p>Must be volatile as it is accessed from various threads. But it's best effort, setting this
    * doesn't have to be atomic.
    */
-  volatile @MonotonicNonNull Throwable subscriptionError;
+  private volatile @MonotonicNonNull Throwable subscriptionError;
+
+  /**
+   * May only ever be altered from within {@link #stop()} or {@link #getNextRecord()} to prevent
+   * race conditions when cancelling subscribers.
+   */
+  private boolean isStopped = false;
```

Comment:
I noticed we have an issue in the errorHandler. If we cancel subscribers from there we risk race conditions as we might concurrently update the state map in `getNextRecord` for resharding events. A better solution is to move all of this into `getNextRecord` and make sure it's only done once using this flag.